### DISCUSSION TITLE
Guide for using CosmWasm ICA contracts

### DISCUSSION THREAD

**Discussion by zanicar** [archway-network/discussions#3] [Smart Contract Development]
We have received a number of questions and requests around Interchain Account (ICA) features and as such would like to open a discussion with out community on this topic.

From an engineering perspective it is imperative that we are cognizant of both the scope of the features that we offer and the cost or technical debt we incur to do so. Our objective is to best serve our community with a protocol that is flexible and easy to use, and balancing these aspects can be a tricky prospect. Thus, in order to assure we deliver features that are valuable to the majority of our developers we would like to experiment with a more interactive approach...

As a starting point we would like our developers to explore the CosmWasm IBC Demo, found [here](https://github.com/confio/cw-ibc-demo/tree/main), with specific focus on the following two cases related to ICA:

- [Simple ICA Contoller](https://github.com/confio/cw-ibc-demo/tree/main/contracts/simple-ica-controller)
- [Simple ICA Host](https://github.com/confio/cw-ibc-demo/tree/main/contracts/simple-ica-host)

We would like you to **discuss your experience** and relay the good, the bad and the ugly...

Our objective with this experiment is to identify how to best serve our community, whether through enhanced documentation, more comprehensive examples, boiler plate code in the form of templates and design patterns, protocol bindings or protocol enhancements.

_We want you to guide us on this journey... as we are on this journey along with you as our community of developers!_

**Comment by zanicar** [comment]
Please note that there is an open issue on CosmWasm for improved native ICA support.
https://github.com/CosmWasm/wasmd/issues/947

**Comment by jlm0** [comment]
@zanicar Thank you for opening this discussion on Interchain Accounts (ICA)!
Not sure if relevant to the feedback you're asking for but I do want to express that the Simple ICA Controller and Host documentation do not satisfy ICS27 for true ICA capabilities. The limitations around requiring CosmWasm contracts on both sides and the inability to connect to non-CosmWasm chains are significant. Since the reflect contract instance holds the account created by the factory it is permanently tied to that factory contract and can't be utilized by other contracts.  Additionally, the flow of messages is restricted to Controller -> Host rather than allowing bidirectional communication. While these Simple ICA contracts are useful examples of message passing, they don't offer the capabilities of chain-agnostic ICA.

I'm curious to know Archway's position on the usage of these Simple ICA contracts. Will there be a push to achieve full ICS27-compatible ICA for Archway?

**Reply by zanicar** [reply]
Yes, this is the feedback we are looking for...

It is clear developers want to progress beyond the current limitations towards native level ICS27 ICA capabilities. However, as per [this issue](https://github.com/CosmWasm/wasmd/issues/947), it may be a while before we get there as an ecosystem. We have some ideas and directions we want to explore and would love to do so with our developer community.

In the light of security concerns we will have to approach this like true ninjas: slowly and cautiously... with both swords drawn! 🥷

**Reply by jlm0** [reply]
@zanicar Thank you for the follow-up!

Given the open issue and unknown timeline for direct ICA message support in wasmd, I'm curious what Archway's stance is here. Is the plan to wait for wasmd implementation before pursuing full ICA on Archway? Or are there ideas and workaround directions you mentioned that could enable more ICA functionality in a nearer term?

Specifically, I'm interested in learning more about the ideas and directions you want to explore with the developer community. Are these approaches that could provide ICA-like capabilities as a stopgap until native support is added to wasmd? It would be great to get more details on what potential options are on the table and their viability.

**Comment by zanicar** [comment]
[Polytone](https://github.com/DA0-DA0/polytone) looks like a great solution to addressing this issue!

cc: @jlm0
